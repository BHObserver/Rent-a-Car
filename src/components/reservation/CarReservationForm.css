* {
  margin: 0;
  padding: 0;
  font-family: 'Gill Sans', 'Gill Sans MT', Calibri, 'Trebuchet MS', sans-serif;
}

.reservation-page {
  margin: 12px;
  padding: 16px;
  width: 100%;
  height: 93vh;
  position: relative;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
}

.reservation-page::before {
  content: '';
  position: absolute;
  width: 100%;
  height: 93vh;
  opacity: 0.2;
  z-index: -1;
}

label {
  color: #fff;
}

.form-container {
  display: flex;
  flex-direction: column;
  width: 100%;
  height: auto;
  padding: 20px;
  border-radius: 8px;
}

.input-container {
  width: 100%;
  gap: 10px;
  height: auto;
}

.input-container input {
  width: 100%;
  height: 40px;
  padding: 12px;
  border: none;
  outline: none;
  border: 1px solid #fff;
  border-radius: 5px;
  box-shadow: 0.3px 0.3px black;
}

.button-container {
  display: flex;
  flex-direction: column;
  justify-content: center;
  width: 400px;
}

.button-container button {
    background-color: #96bf01;
    color: white;
    border: none;
    cursor: pointer;
    border: 1px solid #fff;
    transition: background-color 0.3s ease;
    border-radius: 50px;
    width: 214px;
    padding: 10px;
    font-size: 14px;
    outline: none;
}

.button-container button:hover {
  background-color: #0056b3;
}

.select-container {
  position: relative;
  margin-bottom: 20px;
  width: 214px;
}

.select-container select {
  width: 100%;
  padding: 10px;
  border: 1px solid #77950d;
  font-family: 'Segoe UI', Tahoma, Geneva, Verdana, sans-serif;
  border-radius: 5px;
  appearance: none;
  background-color: #fff;
  cursor: pointer;
  border-radius: 20px;
  color: #77950d;
  outline: none;
}

.select-container select:focus {
  outline: none;
  border-color: #2196f3;
}

.select-arrow {
  position: absolute;
  top: 50%;
  right: 10px;
  transform: translateY(-50%);
  pointer-events: none; /* Ensure the arrow doesn't interfere with the select functionality */
}

.select-arrow::before {
  content: '\\25BC'; /* Unicode character for down arrow */
  font-size: 0.8em;
  color: #888;
}


.error {
  color: red;
  margin-top: 10px;
}

.success {
  color: green;
  margin-top: 10px;
}

@media (min-width: 1024px) {
  .input-container {
    display: flex;
  }
}