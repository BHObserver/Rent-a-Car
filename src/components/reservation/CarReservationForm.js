/* eslint-disable jsx-a11y/label-has-associated-control */
import React, { useState, useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { useNavigate, useParams } from 'react-router-dom';
import { IconButton } from '@mui/material';
import { styled } from '@mui/system';
import { ArrowBack } from '@mui/icons-material';
import { fetchCars } from '../../redux/actions/carActions';
import { createReservation } from '../../redux/actions/reservationActions';
import './CarReservationForm.css';

function CarReservationForm() {
  const [reservedDate, setReservedDate] = useState('');
  const [startTime, setStartTime] = useState('');
  const [endTime, setEndTime] = useState('');
  const [startLocation, setStartLocation] = useState('');
  const [destination, setDestination] = useState('');
  const [totalCost, setTotalCost] = useState('');
  const [selectedCarId, setSelectedCarId] = useState('');
  const [successMessage, setSuccessMessage] = useState('');
  const [failureMessage, setFailureMessage] = useState('');
  const [error, setError] = useState('');
  const [availableCars, setAvailableCars] = useState([]);
  const user = useSelector((state) => state.auth.user);
  const cars = useSelector((state) => state.car.cars.cars) || [];
  const dispatch = useDispatch();
  const navigate = useNavigate();
  const { id } = useParams(); // Get car ID from URL params

  useEffect(() => {
    dispatch(fetchCars());
    // If car ID is provided in the URL, set it as the selected car ID
    if (id) {
      setSelectedCarId(id);
    }
  }, [dispatch, id]);

  const handleSubmit = async (event) => {
    event.preventDefault();

    const validateFields = () => {
      if (!reservedDate
        || !startTime
        || !endTime
        || !startLocation
        || !destination
        || !totalCost
        || !selectedCarId) {
        setError('Please fill in all fields.');
        return false;
      }
      return true;
    };

    if (!validateFields()) {
      return;
    }

    const reservationData = {
      reserved_date: reservedDate,
      start_time: startTime,
      end_time: endTime,
      total_cost: totalCost,
      start_location: startLocation,
      destination,
      user_id: user.accessToken,
      car_id: selectedCarId,
    };

    try {
      await dispatch(createReservation(reservationData));
      setSuccessMessage('Car reserved successfully!');
      setReservedDate('');
      setStartTime('');
      setEndTime('');
      setStartLocation('');
      setDestination('');
      setTotalCost('');
      setError('');
      dispatch(fetchCars());
      const updatedAvailableCars = cars.filter((car) => car.available);
      setAvailableCars(updatedAvailableCars);
      // If a car was selected from the details page, reset the selectedCarId
      setSelectedCarId('');
      navigate('/profile'); // Navigate back to profile after reservation
    } catch (error) {
      setFailureMessage('Error reserving car');
    }
  };

  const handleFindAvailableCars = () => {
    const availableCars = cars.filter((car) => car.available);
    setAvailableCars(availableCars);
  };

  // Show available cars button and car selection field only if no car is selected
  const showCarSelection = !selectedCarId;
  // Check if the user is on the reservation page
  const isOnReservationPage = window.location.pathname === `/reserve/${id}` || '/reserve';

  const StyledIconButtonLeft = styled(IconButton)({
    position: 'absolute',
    top: '30px',
    left: '10px',
    transform: 'translateY(-50%)',
    backgroundColor: '#96bf01',
    border: '1px solid #fff',
    color: 'white',
    width: '60px',
    height: '60px',
    borderRadius: '100% 20% 20% 100%',
    '&:hover': {
      backgroundColor: '#0d47a1',
    },
  });

  return (
    <div className="reservation-page">
      <StyledIconButtonLeft
        aria-label="back to profile"
        onClick={() => navigate('/profile')}
        sx={{ mr: 2 }} // Adjust margin as needed
      >
        <ArrowBack />
      </StyledIconButtonLeft>
      <style>
        {`
        .MuiDrawer-root.MuiDrawer-docked.css-q56gz0-MuiDrawer-docked {
          display: ${isOnReservationPage ? 'none' : 'block'};
        }
        main {
          background-image: linear-gradient(#96bf01ad, #a3cb14bc), url(https://www.travelperk.com/wp-content/uploads/car-rental-companies-1-scaled.jpg);
          background-size: cover;
        }
      `}
      </style>
      <div className="form-container">
        <form onSubmit={handleSubmit} className="reservation-form">
          <div className="input-container">
            <label>
              Reserved Date:
              <input type="date" className="input-field" value={reservedDate} onChange={(e) => setReservedDate(e.target.value)} required />
            </label>
            <label>
              Start Time:
              <input type="datetime-local" className="input-field" value={startTime} onChange={(e) => setStartTime(e.target.value)} required />
            </label>
            <label>
              End Time:
              <input type="datetime-local" className="input-field" value={endTime} onChange={(e) => setEndTime(e.target.value)} required />
            </label>
            <label>
              Start Location:
              <input type="text" className="input-field" value={startLocation} onChange={(e) => setStartLocation(e.target.value)} required />
            </label>
            <label>
              Destination:
              <input type="text" className="input-field" value={destination} onChange={(e) => setDestination(e.target.value)} required />
            </label>
            <label>
              Total Cost:
              <input type="number" className="input-field" value={totalCost} onChange={(e) => setTotalCost(e.target.value)} required />
            </label>
          </div>
          <div className="button-container">
            {showCarSelection && (
            <button type="button" onClick={handleFindAvailableCars} className="form-button">Find Available Cars</button>
            )}
            {showCarSelection && availableCars.length > 0 && (
            <div className="select-container">
              <label>
                Select a Car:
                <select
                  value={selectedCarId}
                  onChange={(e) => setSelectedCarId(e.target.value)}
                  required
                  className="select-car"
                >
                  <option value="">Select a Car</option>
                  {availableCars.map((car) => (
                    <option key={car.id} value={car.id}>
                      {car.make}
                      {' '}
                      {car.model}
                    </option>
                  ))}
                </select>
              </label>
            </div>
            )}
            <button type="submit" className="form-button">Submit</button>
          </div>
        </form>

        {error && <p className="error">{error}</p>}
        {successMessage && <p>{successMessage}</p>}
        {failureMessage && <p>{failureMessage}</p>}
      </div>
    </div>
  );
}

export default CarReservationForm;
