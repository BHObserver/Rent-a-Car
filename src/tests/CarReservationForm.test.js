import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom'; // Import BrowserRouter
import renderer from 'react-test-renderer';
import { useDispatch, useSelector } from 'react-redux';
import CarReservationForm from '../components/reservation/CarReservationForm';

// Mocking useDispatch and useSelector hooks
jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: jest.fn(),
  useSelector: jest.fn(),
}));

describe('CarReservationForm component', () => {
  it('renders correctly', () => {
    // Mocking useDispatch and useSelector hooks
    const dispatch = jest.fn();
    useDispatch.mockReturnValue(dispatch);
    useSelector.mockReturnValue({
      auth: { user: { accessToken: 'mockAccessToken' } },
      car: { cars: { cars: [] } },
    });

    const component = renderer.create(
      <Router>
        {' '}
        {/* Wrap CarReservationForm with BrowserRouter */}
        <CarReservationForm />
      </Router>,
    );
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });
});
