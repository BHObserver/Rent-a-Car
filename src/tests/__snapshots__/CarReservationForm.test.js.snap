// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarReservationForm component renders correctly 1`] = `
Array [
  " ",
  <div
    className="reservation-page"
  >
    <button
      aria-label="back to profile"
      className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1qyg5uq-MuiButtonBase-root-MuiIconButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="ArrowBackIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20z"
        />
      </svg>
    </button>
    <style>
      
        .MuiDrawer-root.MuiDrawer-docked.css-q56gz0-MuiDrawer-docked {
          display: none;
        }
        main {
          background-image: linear-gradient(#96bf01ad, #a3cb14bc), url(https://www.travelperk.com/wp-content/uploads/car-rental-companies-1-scaled.jpg);
          background-size: cover;
        }
      
    </style>
    <div
      className="form-container"
    >
      <form
        className="reservation-form"
        onSubmit={[Function]}
      >
        <div
          className="input-container"
        >
          <label>
            Reserved Date:
            <input
              className="input-field"
              onChange={[Function]}
              required={true}
              type="date"
              value=""
            />
          </label>
          <label>
            Start Time:
            <input
              className="input-field"
              onChange={[Function]}
              required={true}
              type="datetime-local"
              value=""
            />
          </label>
          <label>
            End Time:
            <input
              className="input-field"
              onChange={[Function]}
              required={true}
              type="datetime-local"
              value=""
            />
          </label>
          <label>
            Start Location:
            <input
              className="input-field"
              onChange={[Function]}
              required={true}
              type="text"
              value=""
            />
          </label>
          <label>
            Destination:
            <input
              className="input-field"
              onChange={[Function]}
              required={true}
              type="text"
              value=""
            />
          </label>
          <label>
            Total Cost:
            <input
              className="input-field"
              onChange={[Function]}
              required={true}
              type="number"
              value=""
            />
          </label>
        </div>
        <div
          className="button-container"
        >
          <button
            className="form-button"
            onClick={[Function]}
            type="button"
          >
            Find Available Cars
          </button>
          <button
            className="form-button"
            type="submit"
          >
            Submit
          </button>
        </div>
      </form>
    </div>
  </div>,
]
`;
